src/game_api/game_manager.hpp

```diff
@@ -70,6 +70,43 @@ struct BackgroundMusic
     uint32_t unknown22;
 };
 
+struct KeyboardKey
+{
+    /// Key is being held
+    bool down;
+    size_t unknown;
+};
+
+struct ControllerButton
+{
+    /// Button is being held
+    bool down;
+    /// Button was just pressed down this frame
+    bool pressed;
+};
+
+struct ControllerInput
+{
+    std::array<ControllerButton, 16> buttons;
+};
+
+struct RawInput
+{
+    /// State of all keyboard buttons in a random game order as usual
+    std::array<KeyboardKey, 112> keyboard;
+    /// State of controller buttons per controller. Zero-based indexing, i.e. use game_props.input_index directly to index this.
+    std::array<ControllerInput, 12> controller;
+};
+
+struct SomeInput
+{
+    bool enabled;
+    uint8_t padding1[3];
+    int8_t input_index;
+    uint8_t padding2[3];
+    uint32_t buttons;

```

Comment:
The struct is 
4x bool
int8 input_index
//probably padding
int32 buttons
int8 unknown

Reply:
the first bool is disabled not enabled
second one is like enabled or something (setting false disables it)
dunno about third
fourth is "connected" - if set to false you get popup about controller disconnected

Reply:
ok, third one is like "can_use_menu"

Reply:
Oh yeah I figured there must be more of them, but didn't actually even test what the first one does. The controllers kept getting reserved, not being able to switch to again, even though not set in any input_index, and some of these is probably why

Reply:
The last byte is used as first parameter to `XInputSetState`, at least for the XInput controllers

Reply:
also the buttons are just copy over from `GameProps.buttons`, it's just x4 that, then x4 dupe (for each player), then the last one is for the menu

Reply:
The difference between this and state.inputs buttons i guess is for example: those direct inputs, while the state have the run button held all the time if you have the setting always run enabled

Reply:
Also buttons and buttos_dupe are used to determinate held status, since buttons_dupe are just buttons 1 frame later

Reply:
also also, all of them are really 16bits, not 32, thou not sure if it matters

Reply:
and the menu has to be 16bit since it's buttons_one_frame_later and buttons, and yes, in that order, so it's not nice and easy xd

Reply:
didn't know how to name all
![image](https://github.com/spelunky-fyi/overlunky/assets/36127424/c2824acf-ac50-449d-bec9-3f9cbd4e5af9)

Reply:
Yeah I already have that enum in the api for menu input

Reply:
btw. what `modal_open` suppose to be?
i think is just the player slot used for the icon in pause menu

Reply:
> the first bool is disabled not enabled
> second one is like enabled or something (setting false disables it)
> dunno about third
> fourth is "connected" - if set to false you get popup about controller disconnected

these didn't make any sense after all, everything you said seems backwards

Reply:
> btw. what `modal_open` suppose to be?
> i think is just the player slot used for the icon in pause menu

Yeah I didn't know what it actually does, and that sounds logical, but I just used it to detect if the menu or the journal is open, cause that's when that icon is set too.

Reply:
the fo

> > the first bool is disabled not enabled
> > second one is like enabled or something (setting false disables it)
> > dunno about third
> > fourth is "connected" - if set to false you get popup about controller disconnected
> 
> these didn't make any sense after all, everything you said seems backwards

the fourth i made mistake it's more like "lost_connection" the rest should be correct, so by default you have:
```
false
true
true
false
```

Reply:
Except for keyboard those are not default

Reply:
i guess we're wrong about the first two, and keyboard doesn't seam to care about the menu one
or the struct are different for keyboard vs controller

Reply:
Whatever, not really even smart to expose that struct to the api...

Reply:
also the struct is much larger, seen accessed qword at offset 0x280